feat: append OTP info to output

BumbleFlash opened the pull request:
PR to append the output to the Chicago ride-hailing data, discussed in #11. (WIP)

TODO:
- [x] Review if TncToGtfsWriter should be removed (are we going to use it as-is). Update documentation either way to make it clearer what the parameters do

BumbleFlash committed:
feat: used kotlin flow to convert callback to a suspending function

BumbleFlash commented:
@barbeau This is a working sample using Kotlin flow that converts the callbacks to suspending functions. I'm still figuring out the way to make batches of requests and zip them. In all the examples here, they create `n` variables for `n` requests and then zip them at the end. 

Also, the CI is gonna fail because the library is only in `mavenLocal`

BumbleFlash committed:
chore: convert list to flow

BumbleFlash commented:
Update: @barbeau I pushed a change that is still synchronous but it's way faster than it used to be. I'm transforming the Chicago list to flow using the inbuild filter and collection to remove location values and perform callbacks on the IO thread. 

Here's how I measured the efficiency;-
I measured elapsed time between the start and end of the API callback using the function `getPlanData()` in the first commit 760dce9 and the `makePlanFlow` function in the second commit and here are comparable results. You'll see that converting the list to flow is more consistent and faster than looping it ourselves. Also, I'm calling the API in the `Dispatcher.IO` context and however, the printing still uses the main thread.  


|            | getPlanData() | makePlanFlow () |
| ---------- | ------------- | --------------- |
| Request 1  | 740ms         | 400ms           |
| Request 2  | 23ms          | 25ms            |
| Request 3  | 45ms          | 30ms            |
| Request 4  | 20ms          | 27ms            |
| Request 5  | 89ms          | 16ms            |
| Request 6  | 76ms          | 20ms            |
| Request 7  | 30ms          | 23ms            |
| Request 8  | 57ms          | 19ms            |
| Request 9  | 21ms          | 29ms            |
| Request 10 | 29ms          | 21ms            |

barbeau commented on pom.xml:
```diff
@@ -144,6 +144,22 @@
+            <version>1.4.2</version>
+        </dependency>
+        <dependency>
+            <groupId>io.reactivex.rxjava2</groupId>
```
I think we can get rid of RxJava here?

barbeau commented on src/main/java/edu/usf/cutr/grha/otp/OtpService.kt:
```diff
@@ -0,0 +1,105 @@
+        var start = 0L
+        val planFlow: Flow<Planner> = flow {
+            while (i < 42) {
+                if (chicagoTncData[i].pickupCentroidLatitude == 0.0 ||
```
I'd put this test for valid locations in a private `validLocation()` method to help with readability.

barbeau commented on src/main/java/edu/usf/cutr/grha/otp/OtpService.kt:
```diff
@@ -0,0 +1,105 @@
+        runBlocking {
+            dataFlow
+                .filter {
+                    it.pickupCentroidLatitude != 0.0
```
I see now you'd only using this method because it's faster 🚀 - same on moving check to private method.

barbeau commented:
Nice! Could you add an average value at the bottom of each column? That should be easier to see the differences.

barbeau commented:
Also, typically you'd throw out the first value because of JIT - it will always be much larger than the rest because some of the code get's compiled on the fly.

BumbleFlash committed:
build: removed rxJava dependency

BumbleFlash commented:
@barbeau Apologies, just realized I was feeding two longitude values but it looks like it doesn't affect the processing times although the print takes longer since they actually have data to write.

BumbleFlash committed:
chore: added validation function and fixed latLong values

BumbleFlash committed:
chore: improved planData function

BumbleFlash commented:
@barbeau I'm having trouble emitting multiple values and collecting as one since this is what we're trying to do if the order of the data doesn't matter. However, the Kotlin `flow` manages to work in sync even if I emit the value from a different coroutine scope and collect it from the main. It's like it won't proceed to the next `emit()` before `collect()`

barbeau commented:
>However, the Kotlin flow manages to work in sync even if I emit the value from a different coroutine scope and collect it from the main. It's like it won't proceed to the next emit() before collect()

If I'm understanding correctly, I think that's expected behavior - `emits()` only execute when then are collected. So all the `emits()` before the first `collect()` should execute in parallel on the call to the first `collect()`. But after that, if you try to `emit()` again, it won't execute until that flow is collected.

For producer/consumer problem, have you looked at Channels?

https://kotlinlang.org/docs/channels.html#building-channel-producers

To step back a minute, I think we have three things we're trying to accomplish for out-of-order execution:
1. Parallel execution - All OTP requests and writes to CSV shouldn't block each other (unless required on actual CSV file access to ensure file integrity)
2. Batching execution - We could have X producers (OTP requests) and Y consumers (CSV writes) in parallel

The flow solution at https://stackoverflow.com/questions/60551996/wait-for-result-from-multiple-callbacks-lambdas-in-kotlin/60556171#60556171 is close, but if you're emitting 10 requests and then collecting them, I believe you'll need to wait for all 10 to finish before batching the next 10 (unless you have more than one flow active at a time).

To keep this simple and make some progress towards a working solution, let's just implement the above flow solution first that blocks until the X requests finish, and then we can work on a parallel producer/consumer implementation in another PR.

BumbleFlash commented:
@barbeau Thanks for the explanation. To clarify, we should try to emit multiple calls and collect the response whenever it gets one, irrespective of the order. 

However, I'm having trouble dynamically creating multiple flows and combining them as one to collect them later on the thread.

barbeau commented:
Ok - can you push that code (and comment out or remove other unused code for now) and I can take a look?

BumbleFlash commented:
@barbeau I figured out the solution to this problem finally. It was right in front of me all along. I'm converting the list to a flow and then I'm giving another flow to the `flatMapMerge` function which calls the API. The `flatMapMerge` sequentially executes all the input and collects all the flow concurrently. Guess the solution was a combination of the `getPlanData()` and `makePlanFlow()`. 

```
 runBlocking {
            chicagoTncData
                .asFlow()
                .filter {
                    isValidLocation(
                        it.pickupCentroidLatitude,
                        it.pickupCentroidLongitude,
                        it.dropoffCentroidLatitude,
                        it.dropoffCentroidLongitude
                    )
                }
                .flatMapMerge(concurrency = 10) {
                    flow {
                        val origin = GtfsUtils.latLong(
                            it.pickupCentroidLatitude,
                            it.pickupCentroidLongitude
                        )
                        val destination = GtfsUtils.latLong(
                            it.dropoffCentroidLatitude,
                            it.dropoffCentroidLongitude
                        )

                        val requestParameters = RequestParameters(fromPlace = origin, toPlace = destination)
                        val planApi = PlanApi(url, requestParameters)
                        emit(makePlanRequest(planApi, chicagoTncData.indexOf(it)))
                    }
                }
                .collect {
                    println(it.additionalProperties)
                }
        }
```

The `concurrency` parameter is the maximum number of requests to be executed in parallel. We should able to make it dynamic by passing it in the function signature. 

I tested the concurrency by appending the index of the data to the additional properties map in the `Planner` model and the collection was not in order of the data. So YAY!

BumbleFlash committed:
feat: added concurrency to requests

BumbleFlash committed:
fix: fixed filter condition

BumbleFlash committed:
chore: updated ChicagoTncData.kt model

BumbleFlash committed:
style: added todos for the output algorithm

BumbleFlash commented:
@barbeau I have updated the `ChicagoTncData` model and added TODOS to the `collect` block on how to fill in the output information. Should we merge this PR and open another for the output file?

barbeau commented on src/main/java/edu/usf/cutr/grha/model/ChicagoTncData.kt:
```diff
@@ -38,4 +39,11 @@ data class ChicagoTncData
     @Parsed val pickupCentroidLongitude: Double = 0.0,
     @Parsed val dropoffCentroidLatitude: Double = 0.0,
     @Parsed val dropoffCentroidLongitude: Double = 0.0,
+    @Parsed var totalTravelTime: Int? = 0,
```
I believe we want these metrics for the top 3 trip plan results, IIRC? So for this to output cleanly to CSV I think we'd need to label these "...1", and then have "...2", and "...3" for all the fields.

BumbleFlash replied:
Ahh - I get it. What if there are less than 3 ways to reach a destination? I guess they'll hold the default values then.

BumbleFlash replied:
Quick clarification, should the top 3 be as-is or should we sort by travel time?

barbeau replied:
>Quick clarification, should the top 3 be as-is or should we sort by travel time?

I would leave them sorted as-returned by OTP, because they are ordered by OTP's preference.

We should probably include somehow the priority preference used in the OTP requests, maybe as new model field? QUICK, SAFEST, etc. This obviously impacts the sorting.

BumbleFlash replied:
Right, a common model field for all the 3 trip plan results that shows what the `OptimizeType` was.

barbeau commented on src/main/java/edu/usf/cutr/grha/model/ChicagoTncData.kt:
```diff
@@ -38,4 +39,11 @@ data class ChicagoTncData
+    @Parsed var totalDistance: Double? = 0.0,
+    @Parsed var altitudeChange: Double? = 0.0,
+    @Parsed var transfers: Int? = 0,
+    @Parsed var Modes: List<TraverseModes>? = null
```
Will a `List` flatten to a single CSV record correctly? I think it would need to look like `"BUS,WALK"`.

BumbleFlash replied:
Hmm... Good question. It would be easier to convert it to one `String `then, that way the `ChicagoTncData` class is not dependent on the library as well.

barbeau commented:
>Should we merge this PR and open another for the output file?

I'd say let's do it all in this PR. We probably won't know for sure that the model objects are correct until we actually output it.

BumbleFlash committed:
refactor: added 3 versions of output parameters and changed list to String

BumbleFlash commented:
@barbeau I'm facing a weird issue in OTP v2 here. For some reason, the planner API is not returning any transit information, it only returns the walk paths. Are you able to test it out on your end and see if this is the same case for you? Thanks.

BumbleFlash commented:
I fixed that by updating the GTFS zip to the new one. I'm uploading the newer version to the box folder.

barbeau commented:
>fixed that by updating the GTFS zip to the new one. I'm uploading the newer version to the box folder.

Yes, GTFS expires, so you may need to update occasionally to the newest schedule

barbeau commented:
>>fixed that by updating the GTFS zip to the new one. I'm uploading the newer version to the box folder.
>
>Yes, GTFS expires, so you may need to update occasionally to the newest schedule

@BumbleFlash Actually, thinking more, we should always be using the same GTFS for the same time period, and the OTP requests should include the time and date that the original TNC trip was taken on. We want the transit trip(s) that would have been available on the same date and time that the TNC trip was taken.

You can can download archived GTFS for this time period under the correct date here:
* HART - https://openmobilitydata.org/p/hillsborough-area-regional-transit/228
* CTA - https://openmobilitydata.org/p/chicago-transit-authority/165

BumbleFlash commented:
@barbeau For some reason, the java stack is running out of memory when I use the September CTA GTFS files. Here's the stack trace:-

```
java.lang.OutOfMemoryError: Java heap space
        at java.base/java.util.Arrays.copyOf(Arrays.java:3689)
        at java.base/java.util.ArrayList.grow(ArrayList.java:238)
        at java.base/java.util.ArrayList.grow(ArrayList.java:243)
        at java.base/java.util.ArrayList.add(ArrayList.java:486)
        at java.base/java.util.ArrayList.add(ArrayList.java:499)
        at java.base/java.util.stream.Collectors$$Lambda$70/0x00000001001e1040.accept(Unknown Source)
        at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1675)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at org.opentripplanner.util.MapUtils.mapToList(MapUtils.java:82)
        at org.opentripplanner.gtfs.mapping.StopTimeMapper.map(StopTimeMapper.java:26)
        at org.opentripplanner.gtfs.mapping.GTFSToOtpTransitServiceMapper.map(GTFSToOtpTransitServiceMapper.java:104)
        at org.opentripplanner.gtfs.mapping.GTFSToOtpTransitServiceMapper.mapGtfsDaoToInternalTransitServiceBuilder(GTFSToOtpTransitServiceMapper.java:83)
        at org.opentripplanner.graph_builder.module.GtfsModule.buildGraph(GtfsModule.java:129)
        at org.opentripplanner.graph_builder.GraphBuilder.run(GraphBuilder.java:80)
        at org.opentripplanner.standalone.OTPMain.startOTPServer(OTPMain.java:123)
        at org.opentripplanner.standalone.OTPMain.main(OTPMain.java:39)

```

barbeau commented:
@BumbleFlash Did you try increasing your heap space when running OTP from the command-line? That particular GTFS may have more records for some reason (e.g., it could be a merge between the previous and current period, so it might be twice as big as normal).

BumbleFlash committed:
feat: added date and time options

BumbleFlash committed:
feat: added different time and distance variable for each mode.

BumbleFlash committed:
feat: calculated modeTime and modeDistance for each leg. (WIP)

BumbleFlash commented:
@barbeau Could you please review the above code and suggest a way to reuse the code block code from here https://github.com/CUTR-at-USF/gnss-ride-hailing-analyzer/blob/otp/src/main/java/edu/usf/cutr/grha/otp/OtpService.kt#L101 whenever you have a chance?
 Since they are different variables, I'm not able to create a function for it nor does `Kotlin` support Pass by reference. Thanks!

barbeau commented:
Hmmm...yes, I'm not sure I immediately see a way to improve it. It comes down to needing to flatten out the array of iteraries to a series of variables with the names `...1`, `...2`, etc., and I'm not sure of a less verbose way to do that in Kotlin.

barbeau commented:
I guess it really depends on what the CSV serializer supports. If it serializes arrays automatically then maybe you could store in the model object in an array, and the CSV would take care of writing out to `...1`, `...2`, etc. variables? I don't know if it's that smart...

BumbleFlash commented:
@barbeau Thanks for your insight. I'll try to have a working solution by going down the original path and see if I can optimize it in the future.

BumbleFlash committed:
feat: calculated modeTime and modeDistance for each leg.

BumbleFlash committed:
feat: calculated wait times and added time and distance info to the model object

BumbleFlash committed:
feat: appended travel modes to the model.

BumbleFlash committed:
feat: created CSV writers to output data to a csv file.

BumbleFlash committed:
chore: added commented codes for zoned time objects

BumbleFlash committed:
chore: reduced command line params to remove gps test data

BumbleFlash committed:
refactor: removed GPS test files

BumbleFlash committed:
docs: updated README.md

BumbleFlash committed:
refactored: reordered the chicago model data

BumbleFlash committed:
docs: added time documentation

barbeau committed:
fix: Change to using the CalPoly exported time format

barbeau committed:
chore: Bump Kotlin versions

barbeau committed:
chore: Change to SNAPSHOT version

barbeau committed:
fix: Fix unit tests

barbeau committed:
chore: Remove unrelated functionality

Remove ability to output a GTFS dataset based on TNC data input. This feature is no longer a focus of this project.

barbeau committed:
chore: Fix formatting

barbeau committed:
feat: Add file stream writing, request timeout, try/catch

barbeau committed:
feat: Add socket timeout

And print trip ID for each trip

barbeau committed:
chore: Print URL for each trip too

barbeau committed:
chore: Try and prevent exceptions

Add logging

Also don't fill and return chicagoTncData - just fill, which means it can be a val instead of var

barbeau committed:
chore: Remove/format whitespace

barbeau committed:
Rename .java to .kt

barbeau committed:
feat: Add output file name as 3rd command line parameter

Also convert more files to Kotlin

barbeau committed:
feat: Add Google Transit URL for each output record

barbeau committed:
chore: Suppress the warning for suspicious var

It's needed this was for univocity CSV exporter to work

barbeau committed:
chore: Add units to output, comment out debug output

barbeau commented:
CI is failing because https://github.com/CUTR-at-USF/opentripplanner-client-library/issues/14 hasn't been resolved (that library artifact hasn't been published yet), but I'm going to merge anyway as this is a large PR and the main feature of this application. I'll open an issue separately for getting CI working again.
